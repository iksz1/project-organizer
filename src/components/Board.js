import React, { Component } from "react";
import PropTypes from "prop-types";
import BoardView from "./BoardView/BoardView";
import { observer, inject } from "mobx-react";

@inject("store")
@observer
class Board extends Component {
  static propTypes = {
    store: PropTypes.object,
    match: PropTypes.object
  };

  componentDidMount() {
    const { store, match } = this.props;
    store.fetchData(match.params.boardId);
  }

  render() {
    const { board } = this.props.store;
    return board && <BoardView board={board} />;
  }
}

export default Board;
