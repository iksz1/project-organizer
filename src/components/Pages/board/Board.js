import React, { Component } from "react";
import PropTypes from "prop-types";
import { observer, inject } from "mobx-react";
import BoardView from "./BoardView";

class Board extends Component {
  static propTypes = {
    store: PropTypes.object.isRequired,
    match: PropTypes.object.isRequired
  };

  componentDidMount() {
    const { store, match } = this.props;
    store.boardStore.fetchData(match.params.boardId);
  }

  render() {
    const { boardStore } = this.props.store;

    return <BoardView store={boardStore} />;
  }
}

export default inject("store")(observer(Board));
