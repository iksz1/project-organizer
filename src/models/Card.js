import { types, flow } from "mobx-state-tree";
import db from "../utils/dbWrapper";

export const Card = types
  .model("Card", {
    id: types.number,
    listId: types.number,
    boardId: types.number,
    text: types.string,
    order: types.optional(types.number, 0),
    labelId: types.optional(types.number, 0),
    deleted: types.optional(types.boolean, false)
  })
  .actions(self => ({
    changeText: flow(function* changeText(text) {
      yield db.update("cards", { id: self.id, text });
      self.text = text;
    })
  }));
