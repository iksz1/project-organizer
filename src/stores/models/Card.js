import { types, flow } from "mobx-state-tree";
import db from "../../utils/dbWrapper";

export const CardBase = types.model("CardBase", {
  id: types.number,
  listId: types.number,
  boardId: types.number,
  text: types.string,
  order: types.optional(types.number, 0),
  labelId: types.optional(types.number, 0),
  deleted: types.optional(types.number, 0)
});

const Card = CardBase.named("Card").actions(self => ({
  changeText: flow(function* changeText(text) {
    yield db.update("cards", { id: self.id, text });
    self.text = text;
  })
}));

export default Card;
