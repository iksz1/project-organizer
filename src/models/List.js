import { types, flow, destroy, onPatch } from "mobx-state-tree";
import { Card } from "./Card";
import db from "../utils/dbWrapper";

export const List = types
  .model("List", {
    id: types.number,
    boardId: types.number,
    name: types.string,
    cards: types.optional(types.array(Card), []),
    order: types.optional(types.number, 0),
    deleted: types.optional(types.boolean, false)
  })
  .actions(self => ({
    afterCreate() {
      onPatch(self, snap => console.log(snap));
    },

    changeName: flow(function* changeName(name) {
      const list = yield db.update("lists", { ...self.toJSON(), name, cards: [] });
      self.name = list.name;
    }),

    addCard: flow(function* addCard(text) {
      const card = yield db.add("cards", {
        text,
        listId: self.id,
        boardId: self.boardId,
        order: self.cards.length
      });
      self.cards.push(card);
    }),

    // updateCard: flow(function* updateCard(id, payload) {
    //   const index = self.cards.findIndex(item => item.id === id);
    //   const card = self.cards[index];
    //   self.cards[index] = { ...self.cards[index], ...payload };
    // }),

    deleteCard: flow(function* deleteCard(item) {
      // const index = self.cards.findIndex(item => item.id === id);
      // self.cards.splice(index, 1);
      yield db.remove("cards", item.id);
      destroy(item);
    })
  }));
