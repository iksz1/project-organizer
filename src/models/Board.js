import { types, flow, destroy } from "mobx-state-tree";
import { List } from "./List";
import db from "../utils/dbWrapper";

export const Board = types
  .model("Board", {
    id: types.number,
    name: types.string,
    lists: types.optional(types.array(List), []),
    order: types.optional(types.number, 0),
    deleted: types.optional(types.boolean, false)
  })
  .actions(self => ({
    changeName: flow(function* changeName(name) {
      const board = yield db.update("boards", { ...self.toJSON(), name, lists: [] });
      self.name = board.name;
    }),

    addList: flow(function* addList(name) {
      const list = yield db.add("lists", { boardId: self.id, name, order: self.lists.length });
      self.lists.push(list);
    }),

    deleteList: flow(function* deleteList(item) {
      yield db.remove("lists", item.id);
      destroy(item);
    }),

    moveCard(card, meta) {
      const { fromList, fromIndex, toList, toIndex } = meta;
      if (fromList === toList && fromIndex === toIndex) return; //exit if position hasn't changed

      if (fromList === toList) {
        const list = self.lists.find(item => item.id === fromList);
        const cards = list.cards.slice();
        cards.splice(toIndex, 0, cards.splice(fromIndex, 1)[0]);

        const tmp = cards.slice(fromIndex);
        tmp.forEach((t, i) => (t.order = i + fromIndex));
        cards.splice(fromIndex, tmp.length, ...tmp);

        list.cards = cards;
      } else {
        self.lists[findIndex(self.lists, toList)].cards.splice(toIndex, 0, {
          ...card,
          listId: toList
        });
        self.lists[findIndex(self.lists, fromList)].cards.splice(fromIndex, 1);
      }
    }
  }));

const findIndex = (arr, id) => arr.findIndex(item => item.id === id);
