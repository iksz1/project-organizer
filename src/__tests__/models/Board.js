import { Board } from "../../models/Board";

const boardData = {
  id: 1,
  name: "My new project"
};

it("creates instance successfully", () => {
  const board = Board.create(boardData);
  board.changeName("Project Kangaroo");
  expect(board.toJSON()).toMatchSnapshot();
});

it("can add list", () => {
  const board = Board.create(boardData);
  board.addList("ideas");
  expect(board.toJSON()).toMatchSnapshot();
});

it("can move card", () => {
  const card = { id: 1, listId: 1, boardId: 1, text: "free card" };
  const lists = [
    { id: 1, boardId: 1, name: "first", cards: [card] },
    { id: 2, boardId: 1, name: "second", cards: [] }
  ];
  const meta = { fromList: 1, fromIndex: 0, toList: 2, toIndex: 0 };
  const board = Board.create({ ...boardData, lists });

  board.moveCard(card, meta);
  expect(board.toJSON()).toMatchSnapshot();
});
