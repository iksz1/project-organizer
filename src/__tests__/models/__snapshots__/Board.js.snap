// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can add list 1`] = `
Object {
  "deleted": false,
  "id": 1,
  "lists": Array [
    Object {
      "boardId": 1,
      "cards": Array [],
      "deleted": false,
      "id": 1,
      "name": "ideas",
      "order": 0,
    },
  ],
  "name": "My new project",
  "order": 0,
}
`;

exports[`can move card 1`] = `
Object {
  "deleted": false,
  "id": 1,
  "lists": Array [
    Object {
      "boardId": 1,
      "cards": Array [],
      "deleted": false,
      "id": 1,
      "name": "first",
      "order": 0,
    },
    Object {
      "boardId": 1,
      "cards": Array [
        Object {
          "boardId": 1,
          "deleted": false,
          "id": 1,
          "labelId": 0,
          "listId": 2,
          "order": 0,
          "text": "free card",
        },
      ],
      "deleted": false,
      "id": 2,
      "name": "second",
      "order": 0,
    },
  ],
  "name": "My new project",
  "order": 0,
}
`;

exports[`creates instance successfully 1`] = `
Object {
  "deleted": false,
  "id": 1,
  "lists": Array [],
  "name": "Project Kangaroo",
  "order": 0,
}
`;
